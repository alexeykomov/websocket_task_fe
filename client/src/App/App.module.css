.app {
  text-align: center;
  height: 100%;
}

@media (max-width: 767px) {
  .app {
    height: auto;
  }
}
