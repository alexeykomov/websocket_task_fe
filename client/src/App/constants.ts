export enum MenuState {
  Closed = 'closed',
  Open = 'open',
  Closing = 'closing',
  Opening = 'opening',
}
