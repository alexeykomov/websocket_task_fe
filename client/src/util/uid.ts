let uid = 0;

const premappedIds = new Map();

export const getUid = () => {
  return uid++;
};

export const getUidFor = (reserved: string) => {
  if (!premappedIds.has(reserved)) {
    premappedIds.set(reserved, getUid());
  }
  return premappedIds.get(reserved);
};
