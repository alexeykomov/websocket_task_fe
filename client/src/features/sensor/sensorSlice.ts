import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState, AppThunk } from '../../App/store';
import { ConnectionFilter, PAGE_LENGTH } from './sensorConstants';
import { SetProgressActionPayload } from './sensorTypes';

export interface Sensor {
  id: string;
  name: string;
  connected: boolean;
  progress: boolean;
  unit: string;
  value: string | null;
}

export interface SensorsState {
  sensorIds: string[];
  sensors: Record<string, Sensor>;
  appliedFilter: ConnectionFilter;
  currentPage: number;
}

export const initialState: SensorsState = {
  sensorIds: [],
  sensors: {},
  appliedFilter: ConnectionFilter.All,
  currentPage: 1,
};

type AddSensorDataPayload = {
  id: string;
  name: string;
  connected: boolean;
  unit: string;
  value: string | null;
};

export const sensorSlice = createSlice({
  name: 'sensor',
  initialState: () => initialState,
  reducers: {
    addSensorData: (state, action: PayloadAction<AddSensorDataPayload>) => {
      const id = action.payload.id;
      if (id in state.sensors) {
        state.sensors[id]!.connected = action.payload.connected;
        state.sensors[id]!.value = action.payload.value;
        return;
      }
      const data: Sensor = {
        id,
        name: action.payload.name,
        connected: action.payload.connected,
        value: action.payload.value,
        unit: action.payload.unit,
        progress: false,
      };
      state.sensorIds.push(id);
      state.sensors[id] = data;
    },
    applyFilter: (state, action: PayloadAction<ConnectionFilter>) => {
      state.appliedFilter = action.payload;
      state.currentPage = 1;
    },
    setNextPageInternal: (state) => {
      state.currentPage++;
    },
    setPreviousPageInternal: (state) => {
      state.currentPage--;
    },
    setCurrentPage: (state, action: PayloadAction<number>) => {
      state.currentPage = action.payload;
    },
    setProgress: (state, action: PayloadAction<SetProgressActionPayload>) => {
      if (action.payload.id in state.sensors) {
        state.sensors[action.payload.id]!.progress = action.payload.progress;
      }
    },
  },
});

export const {
  addSensorData,
  applyFilter,
  setCurrentPage,
  setNextPageInternal,
  setPreviousPageInternal,
  setProgress,
} = sensorSlice.actions;

// Thunks.
export const setNextPage = (): AppThunk => (dispatch, getState) => {
  const sensorsState = getState().sensor;
  if (sensorsState.currentPage < selectNumberOfPages(getState())) {
    dispatch(setNextPageInternal());
  }
};

export const setPreviousPage = (): AppThunk => (dispatch, getState) => {
  if (getState().sensor.currentPage > 1) {
    dispatch(setPreviousPageInternal());
  }
};

// Selectors.
export const selectSensors = (state: RootState) =>
  state.sensor.sensorIds
    .map((id) => state.sensor.sensors[id]!)
    .filter((s) => checkFilter(state, s!))
    .map((s) => s.id);

export const checkFilter = (state: RootState, task: Sensor): boolean => {
  switch (state.sensor.appliedFilter) {
    case ConnectionFilter.All:
      return true;
    case ConnectionFilter.Connected:
      return task.connected;
  }
};

export const selectNumberOfPages = (state: RootState) => {
  return Math.ceil(selectSensors(state).length / PAGE_LENGTH);
};

export const selectCurrentPage = (state: RootState) => {
  return state.sensor.currentPage;
};

// Selected sensors for the current page.
export const selectTasksByPage = (state: RootState) => {
  const start = (state.sensor.currentPage - 1) * PAGE_LENGTH;
  const end = start + PAGE_LENGTH;
  return selectSensors(state).slice(start, end);
};

export const selectAppliedFilter = (state: RootState) =>
  state.sensor.appliedFilter;

export const selectSensorData = (id: string) => (state: RootState) =>
  state.sensor.sensors[id]!;

export const sensorReducer = sensorSlice.reducer;
