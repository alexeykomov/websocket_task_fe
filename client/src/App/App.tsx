import React from 'react';
import { Header } from '../components/Header/Header';
import {
  useMenuClosed,
  useSmallScreenLayout,
} from '../hooks/useSmallScreenLayout';
import styles from './App.module.css';
import { MenuState } from './constants';
import { SensorMain } from '../features/sensor/SensorMain';

function App() {
  const smallScreenLayout = useSmallScreenLayout();
  const menuClosed = useMenuClosed();
  const [menuState, setMenuState] = React.useState(
    menuClosed ? MenuState.Closed : MenuState.Open
  );
  const onMenuTransitionEnd = () => {
    if (menuState === MenuState.Closing) {
      setMenuState(MenuState.Closed);
    }
    if (menuState === MenuState.Opening) {
      setMenuState(MenuState.Open);
    }
  };
  const onMenuToggle = () => {
    if (menuState === MenuState.Open) {
      setMenuState(MenuState.Closing);
    }
    if (menuState === MenuState.Closed) {
      setMenuState(MenuState.Opening);
    }
  };
  const onMenuClose = () => {
    if (menuState === MenuState.Open) {
      setMenuState(MenuState.Closing);
    }
  };
  return (
    <div className={styles.app} data-testid="app">
      <Header onMenuToggle={onMenuToggle} fixed={smallScreenLayout} />
      <SensorMain />
    </div>
  );
}

export default App;
