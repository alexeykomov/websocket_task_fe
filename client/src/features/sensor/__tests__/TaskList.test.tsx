import { SensorList } from '../TaskList';
import { renderWithProviders } from '../../../util/test-utils';
import { ConnectionFilter, SensorsState } from '../sensorSlice';
import { TaskListState } from '../../tasklist/taskListSlice';
import { RootState } from '../../../App/store';

let initialStateTask: SensorsState;

let initialStateTaskList: TaskListState;

describe('TaskList', () => {
  beforeEach(() => {
    initialStateTask = {
      sensorIds: [],
      sensors: {},
      currentPage: 1,
      appliedFilter: TaskFilter.All,
    };
    for (let i = 0; i < 20; i++) {
      initialStateTask.sensorIds.push(i);
      initialStateTask.sensors[i] = {
        id: i,
        title: `Task ${i}`,
        completed: false,
        listId: 1,
        createdAt: Date.now(),
        modifiedAt: Date.now(),
      };
    }

    initialStateTaskList = {
      taskListIds: [1],
      taskLists: {
        1: {
          id: 1,
          title: 'Sensor List 1',
          deletable: true,
        },
      },
      activeTaskList: 1,
    };
  });

  it('should render successfully', () => {
    const preloadedState = {
      taskList: initialStateTaskList,
      task: initialStateTask,
    };
    const { baseElement } = renderWithProviders(
      <SensorList />,
      preloadedState as RootState
    );
    expect(baseElement).toBeTruthy();
  });

  it('should render to a snapshot', () => {
    const preloadedState = {
      taskList: initialStateTaskList,
      task: initialStateTask,
    };
    const { baseElement } = renderWithProviders(
      <SensorList />,
      preloadedState as RootState
    );
    expect(baseElement).toMatchSnapshot();
  });
});
