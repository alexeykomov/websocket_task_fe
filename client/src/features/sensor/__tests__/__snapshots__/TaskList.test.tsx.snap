// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should render to a snapshot 1`] = `
<body>
  <div>
    <div
      class="root"
    >
      <input
        class="input"
        placeholder="Add a task"
        type="text"
        value=""
      />
    </div>
    <div
      class="root"
    >
      <div
        class="root"
        data-testid="pager"
      >
        <button
          aria-label="Previous page"
          class="button button-prev"
          disabled=""
          type="button"
        />
        <span
          class="page"
          data-testid="pager-pages"
        >
          1
           /
          2
        </span>
        <button
          aria-label="Next page"
          class="button last button-next"
          type="button"
        />
      </div>
      <div
        class="root"
        data-testid="segmented-control-4"
      >
        <div
          class="label-cont"
        >
          <input
            checked=""
            class="input"
            data-testid="segmented-control-input-4-0"
            id="id-all-4"
            name="segmented-control-4"
            type="radio"
            value="all"
          />
          <label
            class="label"
            data-testid="segmented-control-label-4-0"
            for="id-all-4"
          >
            ALL
          </label>
          <input
            class="input"
            data-testid="segmented-control-input-4-1"
            id="id-active-4"
            name="segmented-control-4"
            type="radio"
            value="active"
          />
          <label
            class="label"
            data-testid="segmented-control-label-4-1"
            for="id-active-4"
          >
            Connected
          </label>
          <input
            class="input"
            data-testid="segmented-control-input-4-2"
            id="id-completed-4"
            name="segmented-control-4"
            type="radio"
            value="completed"
          />
          <label
            class="label label-last"
            data-testid="segmented-control-label-4-2"
            for="id-completed-4"
          >
            Completed
          </label>
        </div>
      </div>
    </div>
    <ul
      class="root"
    >
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 0
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-0"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-0"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 1
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-1"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-1"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 2
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-2"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-2"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 3
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-3"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-3"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 4
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-4"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-4"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 5
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-5"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-5"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 6
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-6"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-6"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 7
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-7"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-7"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 8
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-8"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-8"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
      <li
        class="root"
      >
        <span
          class="textContainer"
        >
          <s />
          <span>
            Sensor 9
          </span>
        </span>
        <input
          class="checkbox"
          id="task-input-9"
          type="checkbox"
        />
        <label
          class="label"
          for="task-input-9"
        >
          <div
            class="indicator"
          />
        </label>
      </li>
    </ul>
  </div>
</body>
`;
